sdk/Pulumi/Provider/PropertyValueSerializer.cs

```diff
@@ -0,0 +1,751 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Pulumi.Serialization;
+
+namespace Pulumi.Experimental.Provider
+{
+    [Obsolete("PropertyValueSerializer is highly experimental, and the shape of the API may be removed or changed at any time. Use at your own risk")]
+    public class PropertyValueSerializer
+    {
+        private object? DeserializeObject(ImmutableDictionary<string, PropertyValue> inputs, Type targetType, string[] path)
+        {
+            var instance = Activator.CreateInstance(targetType);
+            var properties = targetType.GetProperties();
+            foreach (var property in properties)
+            {
+                var propertyName = PropertyName(property);
+                var propertyPath = path.Append(propertyName).ToArray();
+                if (inputs.TryGetValue(propertyName, out var value))
+                {
+                    var deserializedValue = DeserializeValue(value, property.PropertyType, propertyPath);
+                    property.SetValue(instance, deserializedValue);
+                }
+                else
+                {
+                    var maybeDeserializedEmptyValue = DeserializeValue(PropertyValue.Null, property.PropertyType, propertyPath);
+                    // we couldn't find the corresponding property value in the inputs given
+                    // that is alright if the property is optional (i.e. the type is nullable)
+                    // otherwise we throw an error
+                    if (!IsNullable(property.PropertyType) && maybeDeserializedEmptyValue == null)
+                    {
+                        // TODO: implement a proper exception type that includes the type info, errors and property value
+                        var errorPath = "[" + string.Join(", ", propertyPath) + "]";
+                        throw new InvalidOperationException(
+                            $"Could not deserialize object of type {targetType.Name}, missing required property {propertyName} at {errorPath}");
+                    }
+
+                    property.SetValue(instance, maybeDeserializedEmptyValue);
+                }
+            }
+
+            return instance;
+        }
+
+        private bool IsNullable(Type type)
+        {
+            return !type.IsValueType || (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>));
+        }
+
+        private string PropertyName(PropertyInfo property)
+        {
+            var propertyType = property.PropertyType;
+            var propertyName = property.Name;
+            var inputAttribute =
+                propertyType
```

Comment:
Shouldn't this attribute lookup be on the property itself, not the type.

Reply:
Fixed and added tests ✅